File: MATRIX/01_spiral_matrix.cpp
/*
PROBLEM: Spiral Matrix (LeetCode 54)
Difficulty: Medium
Company: Microsoft, Google, Meta, Amazon, Netflix
Pattern: Matrix Traversal

DESCRIPTION:
Given an m x n matrix, return all elements of the matrix in spiral order.

MATRIX DIAGRAM:
  Input Matrix:              Spiral Order:
  1  2  3  4                 1 → 2 → 3 → 4
  5  6  7  8          -->        ↓       ↓
  9  10 11 12                9 → 10→ 11  8
                             ↓           ↑
  Output: [1,2,3,4,8,12,11,10,9,5,6,7]  5 → 6 → 7
  
  Direction: Right → Down → Left → Up (repeat)
  Track boundaries: top, bottom, left, right

APPROACH:
1. Use four pointers: top, bottom, left, right
2. Traverse right: left to right (top row)
3. Traverse down: top to bottom (right column)
4. Traverse left: right to left (bottom row)
5. Traverse up: bottom to top (left column)
6. Shrink boundaries after each traversal

TIME COMPLEXITY: O(M * N)
SPACE COMPLEXITY: O(1) excluding output array
*/

#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        if (matrix.empty()) return {};
        
        vector<int> result;
        int top = 0, bottom = matrix.size() - 1;
        int left = 0, right = matrix[0].size() - 1;
        
        while (top <= bottom && left <= right) {
            // Traverse right
            for (int col = left; col <= right; col++) {
                result.push_back(matrix[top][col]);
            }
            top++;
            
            // Traverse down
            for (int row = top; row <= bottom; row++) {
                result.push_back(matrix[row][right]);
            }
            right--;
            
            // Traverse left (if there's a row remaining)
            if (top <= bottom) {
                for (int col = right; col >= left; col--) {
                    result.push_back(matrix[bottom][col]);
                }
                bottom--;
            }
            
            // Traverse up (if there's a column remaining)
            if (left <= right) {
                for (int row = bottom; row >= top; row--) {
                    result.push_back(matrix[row][left]);
                }
                left++;
            }
        }
        
        return result;
    }
};

int main() {
    Solution sol;
    
    vector<vector<int>> matrix = {
        {1, 2, 3, 4},
        {5, 6, 7, 8},
        {9, 10, 11, 12}
    };
    
    vector<int> result = sol.spiralOrder(matrix);
    
    cout << "Spiral Order: ";
    for (int num : result) {
        cout << num << " ";
    }
    cout << endl;
    
    return 0;
}

/*
INPUT:
matrix = [[1,2,3,4],
          [5,6,7,8],
          [9,10,11,12]]

OUTPUT:
[1,2,3,4,8,12,11,10,9,5,6,7]

EXPLANATION:
Step 1: Traverse right on top row: 1, 2, 3, 4
Step 2: Traverse down on right column: 8, 12
Step 3: Traverse left on bottom row: 11, 10, 9
Step 4: Traverse up on left column: 5
Step 5: Traverse right on remaining: 6, 7
*/

File: MATRIX/01_spiral_matrix.py
"""
PROBLEM: Spiral Matrix (LeetCode 54)
Difficulty: Medium
Company: Microsoft, Google, Meta, Amazon, Netflix
Pattern: Matrix Traversal

DESCRIPTION:
Given an m x n matrix, return all elements of the matrix in spiral order.

APPROACH:
1. Use four pointers: top, bottom, left, right
2. Traverse right: left to right (top row)
3. Traverse down: top to bottom (right column)
4. Traverse left: right to left (bottom row)
5. Traverse up: bottom to top (left column)
6. Shrink boundaries after each traversal

TIME COMPLEXITY: O(M * N)
SPACE COMPLEXITY: O(1) excluding output array
"""

from typing import List


class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        if not matrix:
            return []

        result = []
        top, bottom = 0, len(matrix) - 1
        left, right = 0, len(matrix[0]) - 1

        while top <= bottom and left <= right:
            # Traverse right
            for col in range(left, right + 1):
                result.append(matrix[top][col])
            top += 1

            # Traverse down
            for row in range(top, bottom + 1):
                result.append(matrix[row][right])
            right -= 1

            # Traverse left (if there's a row remaining)
            if top <= bottom:
                for col in range(right, left - 1, -1):
                    result.append(matrix[bottom][col])
                bottom -= 1

            # Traverse up (if there's a column remaining)
            if left <= right:
                for row in range(bottom, top - 1, -1):
                    result.append(matrix[row][left])
                left += 1

        return result


if __name__ == "__main__":
    sol = Solution()

    matrix1 = [[1, 2, 3], [4, 5, 6], [7, 8, 9]]
    print("Spiral order:", sol.spiralOrder(matrix1))  # [1,2,3,6,9,8,7,4,5]

    matrix2 = [[1, 2, 3, 4], [5, 6, 7, 8], [9, 10, 11, 12]]
    print("Spiral order:", sol.spiralOrder(matrix2))  # [1,2,3,4,8,12,11,10,9,5,6,7]
